qi-lib/flow.rkt

```diff
@@ -33,14 +33,24 @@ module, defined after the flow macro. They are all invoked as needed
 in the flow macro.
 |#
 
-(define-syntax-parser flow
-  [(_ onex) ((compose compile-flow expand-flow) #'onex)]
-  ;; a non-flow
-  [_ #'values]
-  ;; error handling catch-all
-  [(_ expr0 expr ...+)
-   (report-syntax-error
-    'flow
-    (syntax->datum #'(expr0 expr ...))
-    "(flow flo)"
-    "flow expects a single flow specification, but it received many.")])
+(define-syntax (flow stx)
+  (syntax-parse stx
+    [(_ onex)
+     #:with name (syntax-local-name)
+     (quasisyntax/loc stx
+       (let ([flowed #,((compose compile-flow expand-flow) #'onex)])
```

Comment:
"flowed" suggests past tense and would be confusing as the name here. `flow` on its own as the name may also be confusing since it could suggest that it is simply a function named "flow", as these two cases would have the same output in the REPL:

```
(define (flow x) x)
flow
```
and
```
(define-flow f (-< add1 add1))
f
```

How about using `compiled-flow` as the name?

I wonder whether in the future it could make sense to do this at a more granular level, so that something like `(object-name (flow (-< ...))) ;=> compiled-tee-flow` and `(object-name (flow (>< ...)) ;=> compiled-amp-flow` -- but only if it could be done without complicating the implementation / incurring performance overhead since otherwise it wouldn't be worth it.

Reply:
> I wonder whether in the future it could make sense to do this at a more granular level, so that something like (object-name (flow (-< ...))) ;=> compiled-tee-flow and (object-name (flow (>< ...)) ;=> compiled-amp-flow -- but only if it could be done without complicating the implementation / incurring performance overhead since otherwise it wouldn't be worth it.

Looks good to me! I've named `amp`, `tee` and `relay` [here](https://github.com/countvajhula/qi/pull/70).

Reply:
Would it be possible to put it in a separate PR from the identity values one? I haven't had a chance to consider the implications of that yet and I'm doubtful that I'll be able to get to it in the near future due to competing priorities. I think we'd be able to merge it sooner if the changes are kept separate (or even included in the present PR). Wdyt?

Reply:
> Would it be possible to put it in a separate PR from the identity values one? I haven't had a chance to consider the implications of that yet and I'm doubtful that I'll be able to get to it in the near future due to competing priorities. I think we'd be able to merge it sooner if the changes are kept separate (or even included in the present PR). Wdyt?

The main job of that PR is to make some forms return the fixed functions in special cases.

Before：
```racket
Welcome to Racket v8.6 [cs].
> (require qi)
> (☯ (-<))
#<procedure:...b/flow/compiler.rkt:108:7>
> (eq? (☯ (-<)) (☯ (-<)))
#f
> (eq? (☯ (-<)) (☯ (gen)))
#f
> (eq? (☯ (-<)) (☯ ⏚))
#f
> (eq? (☯ (-<)) (☯ (select)))
#f
> (☯ (==*))
==*: undefined;
 cannot reference an identifier before its definition
  in module: top-level
 [,bt for context]
> (☯ (switch))
#<procedure:values>
> (☯ (fanout 0))
#<procedure:...b/flow/compiler.rkt:494:9>
> (☯ (fanout 1))
#<procedure:...b/flow/compiler.rkt:494:9>
```

After：
```racket
Welcome to Racket v8.6 [cs].
> (require qi)
> (☯ (-<))
#<procedure:*->1>
> (eq? (☯ (-<)) (☯ (-<)))
#t
> (eq? (☯ (-<)) (☯ (gen)))
#t
> (eq? (☯ (-<)) (☯ ⏚))
#t
> (eq? (☯ (-<)) (☯ (select)))
#t
> (☯ (==*))
#<procedure:1->1>
> (☯ (switch))
#<procedure:values>
> (☯ (fanout 0))
#<procedure:*->1>
> (☯ (fanout 1))
#<procedure:values>
```

Because naming the procedures returned by these forms, and letting these forms return the specific procedures in special cases, are both optimizations to the returned procedures, I thought it would be more convenient to complete these two tasks in one PR. Of course implementing the naming as a PR alone makes sense to me. Wdyt?

Reply:
OK on this. Btw, I will be traveling starting next week but I'll try to get to these before then. I've added a [Calendar](https://github.com/countvajhula/qi/wiki/Calendar) to the wiki to have some visibility into the availability of contributors, as well as other general events of interest that are related to the project (e.g. Michael's upcoming talk at RacketCon is on there, and so is my upcoming talk at EmacsConf).